feat(on-demand IMU calibration): add taskCalibrateImuUsingCore0() oncore 0 and hook to case T_GYRO in reaction()

este-este opened the pull request:
Add function taskCalibrateImuUsingCore0() [imu.h] which uses ulTaskNotifyTake() to await notification of an IMU calibration request.  In imuSetup() [imu.h], run that function as a perpetual task on core 0.  In reaction(), hook to case T_GYRO with C_GYRO_CALIBRATE or/and C1_GYRO_CALIBRATE_IMMEDIATELY (introduces a Character Command sequencing paradigm).  Use xTaskNotifyGive(taskCalibrateImuUsingCore0_handle) to give IMU calibration request notification from the core 1 thread to the core 0 thread.

This allows sending "gc" to get the original calibration behavior but without the need to reboot.  This also allows sending "gci" to get a new "calibrate immediately" behavior for true "on-demand" IMU calibration, again with no reboot.

In OpenCat.h, add documentation regarding Token and Character Commands (including the Character Command sequencing paradigm).  Rename T_CALIBRATE to T_SERVO_CALIBRATE to make its behavior more obvious now that we are calibrating the IMU.

Add a comment ***suggesting*** that C_LEARN be renamed to C_SERVO_FEEDBACK_LEARN and that C_REPLAY be renamed to C_SERVO_FEEDBACK_REPLAY since they appear to only be associated with token T_SERVO_FEEDBACK.

borntoleave reviewed (approved):
Renaming C_arguments accepted. 

For IMU calibration, I added icmQ and mpuQ checks if the macros are defined but the chip is not on the board.

I don't like the calibration task being active all the time, so I moved its call to reaction.h. It now blocks the main tasks (on core1) and the imu reading (on core0) while calibrating, then releases the lock. 
No need to use xTaskNotifyGive.

borntoleave commented:
I cannot push my edits to your pull request. So I merged your branch into main, made my local changes, and pushed it to the repo.

este-este commented:
Hi Rz

Regarding the concern over the "calibration task being active all the
time", I have to disagree but it is your call.  The call to
ulTaskNotifyTake(pdTRUE, portMAX_DELAY) is a blocking call so, until
xTaskNotifyGive is sent, no CPU time is consumed.  See here
<https://sourcevu.sysprogs.com/espressif/lib/freertos/symbols/ulTaskGenericNotifyTake#:~:text=The%20task%20does%20not%20consume%20any%20CPU%20time%20while%20it%20is%20in%20the%20Blocked%20state.>
.

I took a look at your changes and I see that the code specifies the robot
to be placed flat on the floor.  Your code also shuts down the servos.
That is understandable for IMU balance calibration but not for IMU
navigation calibration (my use case).  I envisioned C_GYRO_CALIBRATE to be
the original behavior (flat on the floor, servos off)
and C1_GYRO_CALIBRATE_IMMEDIATELY to be "calibrate right now" in whatever
posture the robot is in.

I do understand that C1_GYRO_CALIBRATE_IMMEDIATELY, as I envisioned, can be
problematic if the operator is relying on IMU balance and the robot is NOT
flat on the floor.  However, that is what the "gc" , i.e. T_GYRO plus
C_GYRO_CALIBRATE is for.  The "gci", i.e. T_GYRO plus C_GYRO_CALIBRATE plus
C1_GYRO_CALIBRATE_IMMEDIATELY is to be only used for IMU navigation.  That
is a long way of saying that I don't think I can use these changes.

Still, with warm regards,

-Tom

On Mon, Apr 21, 2025 at 4:35 AM Rongzhong Li ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> Renaming C_arguments accepted.
>
> For IMU calibration, I added icmQ and mpuQ checks if the macros are
> defined but the chip is not on the board.
>
> I don't like the calibration task being active all the time, so I moved
> its call to reaction.h. It now blocks the main tasks (on core1) and the imu
> reading (on core0) while calibrating, then releases the lock.
> No need to use xTaskNotifyGive.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PetoiCamp/OpenCatEsp32/pull/31#pullrequestreview-2780827299>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDL3NLEQZY2CADEJXRSHIT322SUUHAVCNFSM6AAAAAB2H3SLYKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOOBQHAZDOMRZHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
